src/new/strukt.rs

```diff
@@ -0,0 +1,215 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation of `strukt!`.
+
+/// Constructs a [`New`] that builds up a struct field-by-field, in-place.
+///
+/// This is somewhat different from using `new::of(...).with(...)`, since you
+/// can directly initialize fields using [`New`]s.
+///
+/// ```
+/// # use moveit::{new::New, new, strukt, moveit};
+/// struct Inner {
+///   // ...
+/// }
+///
+/// impl Inner {
+///   pub fn new() -> impl New<Output = Self> {
+///     // ...
+///     # new::of(Self {})
+///   }
+/// }
+///
+/// struct Outer {
+///   x: i32,
+///   y: Inner,
+/// }
+///
+/// impl Outer {
+///   pub fn new() -> impl New<Output = Self> {
+///     strukt!(Self {
+///       x: new::of(42),
+///       y: Inner::new(),
+///     })
+///   }
+/// }
+///
+/// moveit!(let outer = Outer::new());
+/// assert_eq!(outer.x, 42);
+/// ```
+///
+/// If any of the passed-in constructors panics, the destructors of all fields
+/// that were successfully initialized (i.e., those lexically before it) will
+/// run automatically.
+///
+/// # Caveats
+///
+/// The same field cannot be assigned multiple times, like in an ordinary
+/// struct literal.
+///
+/// ```no_compile
+/// # use moveit::{new, strukt};
+/// struct X { a: i32 }
+///
+/// let _ = strukt!(X {
+///   a: new::of(1),
+///   a: new::of(1),
+/// });
+/// ```
+///
+/// Currently, uninferred types and lifetimes cannot occur in the type provided
+/// to `strukt!`. This is a temporary restriction that may be lifted in the
+/// future.
+///
+/// Tuple structs are currently not supported. This is a temporary restriction
+/// that may be lifted in the future.
+#[macro_export]
+macro_rules! strukt {
+  ($strukt:path {$($field:tt $(: $value:expr)?),* $(,)?}) => {{
+    extern crate core as __core;
+    // We cannot follow a $path with a {, so we use an identity type alias
+    // to force rustc to be ok with this. Because $strukt can capture
+    // generic variables from context, we cannot store it in a type alias
+    // directly.
+    type __Id<T> = T;
+
+    // rustc incorrectly thinks that this code causes anything after it
+    // to be unreachable, despite the fact that it is never called!
+    //
+    // This is a closure so that it gets optimized out completely, beyond an
+    // unused function that the linker can delete; otherwise we need to rely on
+    // dead code elimination.
+    #[allow(unused)]
+    let _ = || __Id::<$strukt> {
+      // This verifies that constructing a struct in this way, with dummy
+      // values, is actually valid.
+      $($field: unreachable!()),*
+    };
```

Comment:
> Tuple structs and type inference on the struct type are not well-supported

Is _this_ why inference is not supported? Can't it work if you change the code like this:
```suggestion
    if false {
        #[allow(unused)]
        let _ = || __Id::<$strukt> {
          // This verifies that constructing a struct in this way, with dummy
          // values, is actually valid.
          $($field $(: $value)?),*
        };
    } else {
        /* the rest of the code */
    }
```

if-else should allow you to "move" the values twice (one branch for the static check, the other real), while you can still use a closure to depend less on the dead code elimination...

This should also help with unreachable rustc problem mentioned in the comment above.

Reply:
Nah, that's not why. Type inference isn't supported for the struct type itself because there's no way I could think of to trigger unification between the output type and the struct type in the `offset_of` call. The field types *do* have inference.

Reply:
Oh, hm. Wouldn't a similar `if` help? Like if you unify `this` with `__Id::<$strukt> { $($field $(: $value)?),* };`...